nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md

```diff
@@ -0,0 +1,61 @@
+The routing system can be extended by creating a custom `Feature` and enabling it in an NServiceBus endpoint. To learn more about implementing a routing extension see the [custom routing sample](/samples/routing/custom/)
+
+## Command routing
+
+Routing extensions can access the route table via the feature configuration context:
+
+snippet:RoutingExtensibility-RouteTable
+
+The route table can either be modified once during the feature set up phase or can be passed further e.g. to a `FeatureStartupTask` and updated periodically when the source of routing information changes.
+
+snippet:RoutingExtensibility-StartupTaskRegistration
+
+snippet:RoutingExtensibility-StartupTask
+
+The source parameter is used as a unique key. When the `AddOrReplaceRoutes` is called the first time with a given source, the routes are added to the table. When it is called subsequently, the routes previously registered under that source are replaced by the new routes.
+
+The route table API is thread-safe and atomic, meaning either all the changes from the passed collection or successfully applied or none is.
+
+The routing system prevents route ambiguity. If the new or replaced routes conflict with existing ones, an exception is thrown. It is up to the route extension to deal with that exception but usually it is best practice to trigger the endpoint shutdown to prevent misrouting of messages.
+
+snippet:RoutingExtensibility-TriggerEndpointShutdown 
+
+## Event routing
+
+As mentioned in the [general routing documentation](/nservicebus/messaging/routing.md), event routing differs depending on the transport capabilities. 
+
+Transports supporting the [Publish-Subscribe](/nservicebus/messaging/publish-subscribe/) pattern natively ({include: transports-multicast}) implement the event routing themselves. Refer to specific transport documentation for details on extensibility points.
+
+Transports without that support ({include: transports-unicast}) relay on NServiceBus core routing for event delivery. To emulate multicast routing on top of a unicast transport NServiceBus uses. The key concept is the collection of publishers. For each event it contains information on the logical endpoint that publishes it. Routing extensions can access the publishers collections from the `Feature` set up code:
+
+snippet:RoutingExtensibility-Publishers
+
+The source parameter has the same meaning and effect as in the routes collection.
+
+The publishers collection behaves the same way as the route table with regards to threading and atomicity.
+
+## Physical routing
+
```

Comment:
The structure of the doco is based on the structure of the API (which has four entry points). Not sure how to express that.